test/integration/auth/requestheader_test.go

```diff
@@ -0,0 +1,206 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+	"k8s.io/kubernetes/test/integration/framework"
+	"k8s.io/kubernetes/test/utils/ktesting"
+)
+
+func TestAuthnToKAS(t *testing.T) {
+	tCtx := ktesting.Init(t)
+
+	frontProxyCA, frontProxyClient, frontProxyKey, err := newTestCAWithClient(
+		pkix.Name{
+			CommonName: "test-front-proxy-ca",
+		},
+		big.NewInt(43),
+		pkix.Name{
+			CommonName:   "test-aggregated-apiserver",
+			Organization: []string{"system:masters"},
+		},
+		big.NewInt(86),
+	)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	modifyOpts := func(setUIDHeaders bool) func(opts *options.ServerRunOptions) {
+		return func(opts *options.ServerRunOptions) {
+			opts.GenericServerRunOptions.MaxRequestBodyBytes = 1024 * 1024
+
+			// rewrite the client + request header CA certs with our own content
+			frontProxyCAFilename := opts.Authentication.RequestHeader.ClientCAFile
+			if err := os.WriteFile(frontProxyCAFilename, frontProxyCA, 0644); err != nil {
+				t.Fatal(err)
+			}
+
+			opts.Authentication.RequestHeader.AllowedNames = append(opts.Authentication.RequestHeader.AllowedNames, "test-aggregated-apiserver")
+			if setUIDHeaders {
+				opts.Authentication.RequestHeader.UIDHeaders = []string{"X-Remote-Uid"}
+			}
+		}
+	}
+
+	for _, tt := range []struct {
+		name   string
+		setUID bool
+	}{
+		{
+			name:   "KAS without UID config",
+			setUID: false,
+		},
+		{
+			name:   "KAS with UID config",
+			setUID: true,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			_, kubeConfig, tearDownFn := framework.StartTestServer(tCtx, t, framework.TestServerSetup{
+				ModifyServerRunOptions: modifyOpts(tt.setUID),
+			})
+			defer tearDownFn()
+
+			// Test an aggregated apiserver client (signed by the new front proxy CA) is authorized
+			extensionApiserverClient, err := kubernetes.NewForConfig(&rest.Config{
+				Host: kubeConfig.Host,
+				TLSClientConfig: rest.TLSClientConfig{
+					CAData:     kubeConfig.TLSClientConfig.CAData,
+					CAFile:     kubeConfig.TLSClientConfig.CAFile,
+					ServerName: kubeConfig.TLSClientConfig.ServerName,
+					KeyData:    frontProxyKey,
+					CertData:   frontProxyClient,
+				},
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			selfInfo := &authnv1.SelfSubjectReview{}
+			err = extensionApiserverClient.AuthenticationV1().RESTClient().
+				Post().
+				Resource("selfsubjectreviews").
+				VersionedParams(&metav1.CreateOptions{}, scheme.ParameterCodec).
+				Body(&authnv1.SelfSubjectReview{}).
+				SetHeader("X-Remote-Uid", "test-uid").
+				SetHeader("X-Remote-User", "testuser").
+				SetHeader("X-Remote-Groups", "group1", "group2").
+				Do(tCtx).
+				Into(selfInfo)
+			if err != nil {
+				t.Fatalf("failed to retrieve self-info: %v", err)
+			}
+
+			if selfUID := selfInfo.Status.UserInfo.UID; (len(selfUID) != 0) != tt.setUID {
```

Comment:
nit this is a bit hard to read:

```suggestion
			if hasUID := len(selfInfo.Status.UserInfo.UID) != 0; hasUID != tt.setUID {
```

Reply:
(this nit is outstanding but I cannot tell if you meant to ignore it or forgot about it)